zerver/tests/test_auth_backends.py

```diff
@@ -1471,6 +1476,305 @@ def test_social_auth_invited_as_admin_but_expired(self) -> None:
+        with self.assertRaises(UserProfile.DoesNotExist):
+            get_user_by_delivery_email(old_email, user_profile.realm)
+
+        email_change_object = get_object_from_key(action_key, Confirmation.EMAIL_CHANGE)
+        email_change_object.refresh_from_db()
+        self.assertEqual(email_change_object.status, 1)
+
+    def test_social_auth_email_change_nonexistent_user(self) -> None:
+        '''
+        When user tries to attempt login without an existing account,
+        email change shouldn't happen.
+        '''
+        user_profile = self.example_user('hamlet')
+        new_email = 'hamlet-new@zulip.com'
+        old_email = user_profile.delivery_email
+        account_data_dict = self.get_account_data_dict(email='new-email@zulip.com',
+                                                       name='New King')
+        action_key = create_email_change_confirmation_key(user_profile, new_email)
+
+        self.login_user(user_profile)
+
+        expect_choose_email_screen = False
+        if self.backend == GitHubAuthBackend:
+            # github takes to choose email screen.
+            expect_choose_email_screen = True
+
+        result = self.social_auth_test(account_data_dict, subdomain='zulip',
+                                       action_key=action_key,
+                                       expect_choose_email_screen=expect_choose_email_screen)
+
+        result = self.client_get(result.url)
+        self.assertEqual(result.status_code, 400)
+        self.assert_in_response("Incorrect login. Login with your old email to change your email address",
+                                result)
+
+        with self.assertRaises(UserProfile.DoesNotExist):
+            get_user_by_delivery_email(new_email, user_profile.realm)
+        # No exception should be raised.
+        get_user_by_delivery_email(old_email, user_profile.realm)
+
+    def test_social_auth_email_change_login_through_different_account(self) -> None:
+        user_profile = self.example_user('hamlet')
+        new_email = 'hamlet-new@zulip.com'
+        old_email = user_profile.delivery_email
+        account_data_dict = self.get_account_data_dict(email='iago@zulip.com',
+                                                       name='Admin Iago')
+        action_key = create_email_change_confirmation_key(user_profile, new_email)
+
+        self.login_user(user_profile)
+        result = self.social_auth_test(account_data_dict, subdomain='zulip',
+                                       action_key=action_key)
+
+        result = self.client_get(result.url)
+        self.assertEqual(result.status_code, 400)
+        self.assert_in_response("Incorrect login. Login with your old email to change your email address",
+                                result)
+
+        with self.assertRaises(UserProfile.DoesNotExist):
+            get_user_by_delivery_email(new_email, user_profile.realm)
+        # No exception should raise.
```

Comment:
```suggestion
        # No exception should be raised.
```

Though I'm not sure if that kind of comment isn't rather redundant :thinking: We can clearly see from the code that exceptions aren't expected to happen here :grinning:

Reply:
Thought no one should wonder why's there `get_user_by_delivery_email(old_email, user_profile.realm)` that isn't being used in any way. But yes, it seems to be a bit obvious seeing the two lines just above this and in the context of tests it is expected as far as I can see. So, removed the comments 😀